### ISSUE TITLE
Sound stop after a while when looping

### ISSUE THREAD

**Issue by retameur** [google/ExoPlayer#4913]
### Issue description

Using ExoPlayer 2.8.2 to play a local .m4a sound file in repeat mode, the sound stop after a while (typically a couple of hours). If playing several files simultaneously, each file stop at a different time. No error is reported in neither the player event listener nor the media source event listener.

The error occurs with both `player.setRepeatMode(Player.REPEAT_MODE_ALL);` or by using a LoopingMediaSource object.

When the audio stop, the following lines appears in Logcat:
    1536-1577/? I/AudioFlinger: BUFFER TIMEOUT: remove(4096) from active list on thread 0xf1203680
    1430-1578/? W/audio_hw_generic: Not supplying enough data to HAL, expected position 942828484 , only wrote 934428960

toggling player.setPlayWhenReady(...) does not restore the playback.

We didn't notice this issue (and none of our users complain) with the previous version of ExoPlayer we were using: r2.4.1. Unfortunately it is not possible to roll back as an audio glitch was audible at the end of each loop.

### Reproduction steps

The following code is running in a service:

```
...
dataSourceFactory = new DefaultDataSourceFactory(context, Util.getUserAgent(context, "XXXX"), null);
player = ExoPlayerFactory.newSimpleInstance(context, new DefaultTrackSelector());
MediaSource audioSource = new ExtractorMediaSource.Factory(dataSourceFactory).createMediaSource(audioItem.getUri());
player.setRepeatMode(Player.REPEAT_MODE_ALL);
player.prepare(audioSource);
...
player.setPlayWhenReady(true);
```

Wait until the sound stops

### Link to test content
Sent by email

### Version of ExoPlayer being used
2.8.2

### Device(s) and version(s) of Android being used
Pixel XL simulator with Android Oreo 8.1
Some of our users complained about the same issue on different devices.

### A full bug report captured from the device
Sent by email

**Comment by andrewlewis** [comment]
I couldn't see any obvious issues with the media or in the bug report.

Please could you try to reproduce this in the ExoPlayer demo app, but modified to set repeat mode, and on a physical device rather than using the emulator? If you can reproduce it please send the bug report. If not, that suggests there's something wrong with the service (perhaps a leak or not holding locks required to keep the device awake). Thanks.

**Comment by andrewlewis** [comment]
I think the issue here is that there's a discrepancy between the duration declared in the MP4 container metadata and the actual duration of audio after applying the iTunSMPB trimming for gapless playback. I think the actual duration is 18955260 us (this is the declared duration 19017142 us, which is the duration of audio before trimming, minus the duration of 2112 + 617 audio frames at 44.1 kHz, which are trimmed). This discrepancy causes the renderer position to drift slowly away from the expected position based on the durations of earlier loops, until the player stops loading because it permanently thinks it has enough data buffered, at which point audio runs out.

If you have control over the media you may be able to fix the issue by rewriting the duration in the container to take into account trimmed frames, or possibly switching to using edit lists to signal gapless trimming.

The next steps here are to verify the hypothesis by hardcoding the calculated duration and verifying that playback doesn't get stuck, then possibly altering the extractor to update the duration to compensate for audio frames trimmed based on iTunSMPB metadata.

**Comment by retameur** [comment]
I have been able to reproduce the bug with ExoPlayer Demo 2.8.4 (report sent by email). The playback stops after the same number of loops (which varies with each specific sound file) as in our app. Its 241 for the frogs.m4a file I've sent by email.

We have control over the media, but it is not a solution for us, as the same media are also used for iOS and the Web. And most of the current users have already downloaded the files.

I'm not sure it is related to the iTunSMPB metatag. I've re-encoded the file twice using fdkaac:
    - once with the --gapless-mode option set to 0 (iTunSMPB)
    - once with the option set to 1: ISO standard (edts and sgpd)
The playback of both files fails simultaneously.

As for the steps you are suggesting verifying the hypothesis, I have no idea how to do so.

**Comment by retameur** [comment]
I have also tested with ExoPlayer demo 2.9.0 and the bug seems to be fixed now (if it still occurs, it is after way more loops). So problem solved. 

Thanks for your help.

**Comment by tonihei** [comment]
Yes, it was fixed by a change in 2.9.0 which prevents the problem you describe. Still, if you want perfect gapless playback of the looped file, you should try to ensure the duration exactly matches the content.
